posts/inside-rust/2020-11-02-Using-rustc_cranelift_codegen.md

```diff
@@ -0,0 +1,148 @@
+good at producing fast, highly optimized code, but is not very good at
+compiling code quickly. `cg_clif`, which uses the [Cranelift] project, would
+provide a fast backend which greatly improves compile times, at the cost of
+performing very few optimizations. This is a great fit for debug builds, and the hope is
+that `cg_clif` will eventually be the default backend in debug mode.
+
+## What is the progress of using `rustc_codegen_cranelift` for debug builds?
+
+There has been a [Major Change Proposal][MCP] open for some time for making
+`cg_clif` part of the main Rust repository. Recently, [the MCP was
+accepted][compiler-team#270] and the compiler team [merged][#77975]
+`rustc_cranelift_codegen` [into the main Rust git repository][#77975].
+`cg_clif` is not yet distributed with `rustup`, but this means you can now
+build it from source in-tree!
+
+## How do I use `rustc_codegen_cranelift`?
+
+In this section, I'll walk through step-by-step how to build the new backend from source, then use it on your own projects. All code is copy/paste-able, and each step is explained.
+
+First, let's build `cg_clif` from source.
+
+```sh
+$ git clone https://github.com/bjorn3/rustc_codegen_cranelift.git
+$ ./prepare.sh
+$ cargo build --release
+$ ./build_sysroot/build_sysroot.sh
+```
+
+Now, we can start using it to compile a project. For demonstration purposes,
+I'll be be using `cargo`, but you can use any Rust project supported by
+`cg_clif`.
+
+```
+$ cd ..
+$ git clone https://github.com/rust-lang/cargo/
+$ cd cargo
+$ ../cargo.sh build
+...
+    Finished dev [unoptimized + debuginfo] target(s) in 49.93s
+```
+
+It works! For comparison, let's see how long the equivalent LLVM backend would
+take.
+
+```sh
+$ rustup install nightly-2020-10-31
+$ cargo +nightly-2020-10-31 build
+...
+    Finished dev [unoptimized + debuginfo] target(s) in 54.64s
+```
+
+LLVM takes a full 5 seconds longer for a full build. Next, let's try incremental builds:
+
+```
+$ git apply <<EOF
+diff --git a/src/cargo/lib.rs b/src/cargo/lib.rs
+index bccb41121..703afa754 100644
+--- a/src/cargo/lib.rs
++++ b/src/cargo/lib.rs
+@@ -36,8 +36,8 @@ use anyhow::Error;
+ use log::debug;
+ use std::fmt;
+ 
+-pub use crate::util::errors::{InternalError, VerboseError};
+ pub use crate::util::{CargoResult, CliError, CliResult, Config};
++pub use crate::util::errors::{InternalError, VerboseError};
+ 
+ pub const CARGO_ENV: &str = "CARGO";
+EOF
+$ ../rustc_codegen_cranelift/build.sh
+    Finished dev [unoptimized + debuginfo] target(s) in 7.98s
+$ cargo +nightly-2020-10-31 build
+   Compiling cargo v0.50.0 (/home/joshua/cargo)
+    Finished dev [unoptimized + debuginfo] target(s) in 5.48s
+```
+
+LLVM is actually *faster* here: `cg_clif` doesn't support internal parallelism
+until parallel rustc lands. This means that `cg_llvm` can normally utilize the
+CPU cores more effectively when there are few crates that can be built at a
+given time.
```

Comment:
In this case it turned out to be serde_derive being slower at runtime when compiled using cg_clif than cg_llvm.

Reply:
Updated it with the info about serde_derive.